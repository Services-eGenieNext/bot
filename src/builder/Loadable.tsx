/*
 * Copyright (c) 2020-present Lawly AB. All Rights Reserved.
 * Author: Mehdi Tirgar <mehdi.tirgar@wearelevels.com> 2020-2021
 *
 */

/**
 * Asynchronously loads the component for BotBuilder
 */

import { lazyLoad } from 'shared/utils/loadable';

export const BotBuilder = lazyLoad(
  () =>
    import(
      /* webpackChunkName: "builder" */
      './index'
    ),
  module => module.BotBuilder,
);
