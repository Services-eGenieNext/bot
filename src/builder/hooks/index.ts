/*
 * Copyright (c) 2020-present Lawly AB. All Rights Reserved.
 * Author: Mehdi Tirgar <mehdi.tirgar@wearelevels.com> 2020-2021
 *
 */

export * from 'builder/hooks/useActionHandlers';
export * from 'builder/hooks/useBlockParagraphs';
export * from 'builder/hooks/useCenterBlock';
export * from 'builder/hooks/useConnections';
export * from 'builder/hooks/useElementsUpdater';
export * from 'builder/hooks/useEventHandlers';
export * from 'builder/hooks/useFlowState';
export * from 'builder/hooks/useNewBlockPosition';
export * from 'builder/hooks/useOptionsHandlers';
export * from 'builder/hooks/usePanelParagraphs';
export * from 'builder/hooks/useVariablesData';
