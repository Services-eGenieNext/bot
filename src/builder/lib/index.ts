/*
 * Copyright (c) 2020-present Lawly AB. All Rights Reserved.
 * Author: Mehdi Tirgar <mehdi.tirgar@wearelevels.com> 2020-2021
 *
 */

export * from 'builder/lib/blocks';
export * from 'builder/lib/bots';
export * from 'builder/lib/colors';
export * from 'builder/lib/common';
export * from 'builder/lib/connections';
export * from 'builder/lib/elements';
export * from 'builder/lib/state';
export * from 'builder/lib/validation';
