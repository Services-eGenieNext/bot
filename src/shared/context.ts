/*
 * Copyright (c) 2020-present Lawly AB. All Rights Reserved.
 * Author: Mehdi Tirgar <mehdi.tirgar@wearelevels.com> 2020-2021
 *
 */

import React, { useContext } from 'react';
import mitt from 'mitt';
import { ModuleContextType } from './types';

const moduleContextValue: ModuleContextType = {
  emitter: mitt(),
  appName: 'app',
  registerAction: () => void 0,
  mountNode: document.body,
  options: {
    debug: true,
    language: 'en',
    bots: [],
    updateInterval: 1000,
    userVariables: [],
    globalVariables: [],
    apiTemplates: [],
    variableValues: {
      user: {},
      global: {},
    },
    responses: [],
  },
};

export const ModuleContext =
  React.createContext<ModuleContextType>(moduleContextValue);

export const useModuleContext = () => {
  return useContext(ModuleContext);
};

export const setContext = (
  moduleContext: ModuleContextType,
): ModuleContextType => {
  return Object.assign(moduleContextValue, {
    ...moduleContext,
    options: {
      ...moduleContextValue.options,
      ...moduleContext.options,
    },
  });
};

export const getContext = (): ModuleContextType => {
  return moduleContextValue;
};
