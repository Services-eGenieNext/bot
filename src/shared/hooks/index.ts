/*
 * Copyright (c) 2020-present Lawly AB. All Rights Reserved.
 * Author: Mehdi Tirgar <mehdi.tirgar@wearelevels.com> 2020-2021
 *
 */

export * from './useLocals';
