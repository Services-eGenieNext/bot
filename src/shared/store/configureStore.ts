/*
 * Copyright (c) 2020-present Lawly AB. All Rights Reserved.
 * Author: Mehdi Tirgar <mehdi.tirgar@wearelevels.com> 2020-2021
 *
 */

import { configureStore } from '@reduxjs/toolkit';
import { createLogger } from 'redux-logger';
import { createInjectorsEnhancer, forceReducerReload } from 'redux-injectors';
import createSagaMiddleware from 'redux-saga';
import { createReducer } from './reducers';

const isDevEnv = process.env.NODE_ENV !== 'production';
const reduxLogger = createLogger({ level: 'debug', collapsed: true });

export function configureAppStore() {
  const reduxSagaMonitorOptions = {};
  const sagaMiddleware = createSagaMiddleware(reduxSagaMonitorOptions);
  const { run: runSaga } = sagaMiddleware;

  // Create the store with saga middleware
  const middlewares = [sagaMiddleware];

  if (isDevEnv) {
    middlewares.push(reduxLogger);
  }

  const enhancers = [
    createInjectorsEnhancer({
      createReducer,
      runSaga,
    }),
  ];

  const store = configureStore({
    reducer: createReducer(),
    middleware: getDefaultMiddleware =>
      getDefaultMiddleware().concat(middlewares),
    devTools: isDevEnv && {
      name: 'Bot Builder Store',
      latency: 200,
      autoPause: true,
      shouldRecordChanges: false,
    },
    enhancers,
  });

  // Make reducers hot reloadable, see http://mxs.is/googmo
  /* istanbul ignore next */
  if (module.hot) {
    module.hot.accept('./reducers', () => {
      forceReducerReload(store);
    });
  }

  return store;
}
