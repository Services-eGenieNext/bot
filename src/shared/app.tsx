/*
 * Copyright (c) 2020-present Lawly AB. All Rights Reserved.
 * Author: Mehdi Tirgar <mehdi.tirgar@wearelevels.com> 2020-2021
 *
 */

import React, { FC } from 'react';
import ReactDOM from 'react-dom';
import { Provider } from 'react-redux';
import { Emitter } from 'mitt';
import {
  ConnectedAppProps,
  DestroyAppFunc,
  ModuleContextType,
  RenderAppFunc,
} from 'shared/types';
import { logger } from 'shared/utils';
import { ErrorBoundary } from 'shared/containers/ErrorBoundary';
import { configureAppStore } from 'shared/store/configureStore';
import { ModuleContext, setContext } from './context';
import 'shared/locales/i18n';
import 'react-app-polyfill/stable';
import { clearPreferredLanguageFromLS } from './locales/helper';

const ConnectedApp: FC<ConnectedAppProps> = ({ Component, moduleContext }) => {
  const contextValue = setContext(moduleContext);
  return (
    <ModuleContext.Provider value={contextValue}>
      <Provider store={store}>
        <React.StrictMode>
          <ErrorBoundary>
            <Component />
          </ErrorBoundary>
        </React.StrictMode>
      </Provider>
    </ModuleContext.Provider>
  );
};

export const store = configureAppStore();

export const getApp = (moduleContext: ModuleContextType) => {
  const { appName } = moduleContext;
  let AppComponent: FC = () => null;
  switch (appName) {
    case 'BotBuilder':
      AppComponent = require('builder/Loadable').BotBuilder;
      break;
    case 'BotUi':
      AppComponent = require('user-ui/Loadable').BotUi;
      break;
  }

  return (
    <ConnectedApp Component={AppComponent} moduleContext={moduleContext} />
  );
};

const destroyApp: DestroyAppFunc = (
  mountNode: HTMLElement,
  emitter: Emitter,
) => {
  const unmounted = ReactDOM.unmountComponentAtNode(mountNode);
  if (unmounted) {
    emitter.emit('destroyed');
  }
  clearPreferredLanguageFromLS();
};

export const renderApp: RenderAppFunc = moduleContext => {
  const { mountNode } = moduleContext;
  if (module.hot) {
    module.hot.accept(['builder/Loadable', './locales/i18n'], () => {
      ReactDOM.hydrate(getApp(moduleContext), mountNode);
    });
  }

  const App = () => getApp(moduleContext);
  if (mountNode.childElementCount > 0) {
    logger.warn(
      'The mount node contains child elements, hydrating instead of rendering.',
    );
    ReactDOM.hydrate(<App />, mountNode);
  } else {
    ReactDOM.render(<App />, mountNode);
  }

  return destroyApp;
};
