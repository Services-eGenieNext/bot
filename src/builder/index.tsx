/*
 * Copyright (c) 2020-present Lawly AB. All Rights Reserved.
 * Author: Mehdi Tirgar <mehdi.tirgar@wearelevels.com> 2020-2021
 *
 */

import React, { useEffect } from 'react';
import { StylesProvider } from '@material-ui/core';
import { ThemeProvider } from '@material-ui/core/styles';
import theme from 'shared/styles/theme';
import { useModuleContext } from 'shared/context';
import {
  sanitizeBotsList,
  createGenerateClassName,
  useInjectReducer,
  useInjectSaga,
  logger,
} from 'shared/utils';
import BotEditor from 'builder/containers/BotEditor';
import { botEditorSaga } from 'builder/features/saga';
import { reducer, sliceKey } from 'builder/features/slice';
import './fonts.css';

const generateClassName = createGenerateClassName({
  productionPrefix: 'lawly',
});

export function BotBuilder() {
  useInjectReducer({ key: sliceKey, reducer: reducer });
  useInjectSaga({ key: sliceKey, saga: botEditorSaga });
  const { emitter, options } = useModuleContext();

  useEffect(() => {
    return () => {
      emitter.emit('unloaded');
      emitter.all.set('unloaded', []);
    };
  }, [emitter]);

  const hasBots =
    options.bots &&
    options.bots.length > 0 &&
    Object.keys(sanitizeBotsList(options.bots)).length;

  if (!hasBots) {
    logger.error(
      'BotBuilder has been initialized with empty or invalid bots, therefore not rendering',
    );
    return null;
  }

  return (
    <ThemeProvider theme={theme}>
      <StylesProvider generateClassName={generateClassName}>
        <BotEditor />
      </StylesProvider>
    </ThemeProvider>
  );
}
