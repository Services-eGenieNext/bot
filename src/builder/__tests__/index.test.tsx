/*
 * Copyright (c) 2020-present Lawly AB. All Rights Reserved.
 * Author: Mehdi Tirgar <mehdi.tirgar@wearelevels.com> 2020-2021
 *
 */

import React from 'react';
import { createRenderer } from 'react-test-renderer/shallow';
import { getApp } from 'shared/app';
import { getContext } from 'shared/context';
import { configureAppStore } from 'shared/store/configureStore';
import { BotBuilder } from 'builder';

const renderer = createRenderer();
const store = configureAppStore();

describe('<App />', () => {
  let realUseContext;
  let useContextMock;

  beforeEach(() => {
    realUseContext = React.useContext;
    useContextMock = React.useContext = jest.fn();
  });

  afterEach(() => {
    React.useContext = realUseContext;
  });

  beforeAll(() => {
    jest.mock('app/builder/Loadable', () => require('builder/index'));
  });

  it('getApp should render BotBuilder using get app', async () => {
    renderer.render(getApp({ ...getContext(), appName: 'BotBuilder' }));
    const renderedOutput = renderer.getRenderOutput();
    expect(renderedOutput.props.children).toMatchSnapshot();
  });

  it('getApp should render BotBuilder directly', async () => {
    useContextMock.mockReturnValue({ store });
    renderer.render(<BotBuilder />);

    const renderedOutput = renderer.getRenderOutput();
    expect(renderedOutput.props.children).toMatchSnapshot();
  });
});
