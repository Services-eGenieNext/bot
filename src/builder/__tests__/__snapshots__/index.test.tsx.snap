// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> getApp should render BotBuilder directly 1`] = `
<StylesProvider
  generateClassName={[Function]}
>
  <BotEditor />
</StylesProvider>
`;

exports[`<App /> getApp should render BotBuilder using get app 1`] = `
<Provider
  store={
    Object {
      "createReducer": [Function],
      "dispatch": [Function],
      "getState": [Function],
      "injectedReducers": Object {},
      "injectedSagas": Object {},
      "replaceReducer": [Function],
      "runSaga": [Function],
      "subscribe": [Function],
      Symbol(Symbol.observable): [Function],
    }
  }
>
  <UNDEFINED>
    <ErrorBoundary>
      <BotBuilder />
    </ErrorBoundary>
  </UNDEFINED>
</Provider>
`;
